test/test_torch.py

```diff
@@ -5302,6 +5297,138 @@ def add_neg_dim_tests():
 
+    def test_assertEqual_debug(self, device):
+        # Scalar comparison
+        s = ".+Comparing 4.0 and 7.0 gives a difference of 3.0.+"
+        with self.assertRaisesRegex(AssertionError, s):
+            self.assertEqual(4., 7.)
+
+        # Scalar complex comparison, real part
+        s = ".+Comparing the real part 1.0 and 3.0 gives a difference of 2.0.+"
+        with self.assertRaisesRegex(AssertionError, s):
+            self.assertEqual(complex(1, 3), complex(3, 1))
+
+        # Scalar complex comparison, imaginary part
+        s = ".+Comparing the imaginary part 3.0 and 5.5 gives a difference of 2.5.+"
+        with self.assertRaisesRegex(AssertionError, s):
+            self.assertEqual(complex(1, 3), complex(1, 5.5))
+
+    @unittest.skipIf(not TEST_NUMPY, 'NumPy not found')
+    @dtypes(*list(product(torch_to_numpy_dtype_dict.keys(),
+                          torch_to_numpy_dtype_dict.keys())))
+    def test_compare_type(self, device, dtypes):
+        a = torch.empty(0, device=device, dtype=dtypes[0])
+        b = torch.empty(0, device=device, dtype=dtypes[1])
+        torch_compare_dtype = self.compare_type(a, b, allow_ints=True)
+        np_compare_dtype = np.promote_types(
+            torch_to_numpy_dtype_dict[dtypes[0]],
+            torch_to_numpy_dtype_dict[dtypes[1]])
+
+        # Note: NumPy returns instances of dtypes so this converts them
+        # to their more commonly used representation
+        # (ex. dtype('int16') -> np.int16)
+        np_compare_dtype = getattr(np, str(np_compare_dtype))
+
+        # Special-cases float16 inputs on non-CUDA devices
+        if np_compare_dtype is np.float16 and self.device_type != 'cuda':
+            self.assertTrue(torch_compare_dtype is torch.float32)
+            return
+
+        # Checks that float16 and bfloat16 have the same promotion on CPU
+        if self.device_type == 'cpu' and dtypes[0] is torch.float16:
+            c = torch.empty(0, device=device, dtype=torch.bfloat16)
+            bfloat16_type = self.compare_type(c, b, allow_ints=True)
+            self.assertTrue(torch_compare_dtype is bfloat16_type)
+
+        self.assertTrue(torch_compare_dtype ==
+                        numpy_to_torch_dtype_dict[np_compare_dtype])
+
+    # Checks that compareTensors provides the correct debug info
+    def test_comparetensors_debug(self, device):
+        # Checks float tensor comparisons (2D tensor)
+        a = torch.tensor(((0, 6), (7, 9)), device=device, dtype=torch.float32)
+        b = torch.tensor(((0, 7), (7, 22)), device=device, dtype=torch.float32)
+        result, debug_msg = self.compareTensors(a, b)
+        expected_msg = ("With rtol=1.3e-06 and atol=1e-05, found 2 element(s) "
+                        "whose difference(s) exceeded the margin of error. "
+                        "The greatest difference was 13.0 (9.0 vs. 22.0), "
+                        "which occurred at index (1, 1).")
+        self.assertTrue(debug_msg == expected_msg)
```

Comment:
nit: a bit nicer to use assertEqual for a better failure message if this changes

Reply:
Sure. Sounds good.